libcalico-go/lib/backend/k8s/conversion/conversion.go

```diff
@@ -273,10 +275,403 @@ func (c converter) ServiceToKVP(service *kapiv1.Service) (*model.KVPair, error)
+			protocolPorts[pStr] = append(protocolPorts[pStr], *calicoPort)
+		}
+	}
+
+	protocols := make([]string, 0, len(protocolPorts))
+	for k := range protocolPorts {
+		protocols = append(protocols, k)
+	}
+	// Ensure deterministic output
+	sort.Strings(protocols)
+
+	// Combine destinations with sources to generate rules. We generate one rule per protocol,
+	// with each rule containing all the allowed ports.
+	for _, protocolStr := range protocols {
+		calicoPorts := protocolPorts[protocolStr]
+		calicoPorts = SimplifyPorts(calicoPorts)
+
+		var protocol *numorstring.Protocol
+		if protocolStr != "" {
+			p := numorstring.ProtocolFromString(protocolStr)
+			protocol = &p
+		}
+
+		var selector, nsSelector string
+		for _, peer := range peers {
+			if peer.Namespaces != nil {
+				selector = ""
+				nsSelector = c.k8sSelectorToCalico(peer.Namespaces, SelectorNamespace)
+			}
+			if peer.Pods != nil {
+				selector = c.k8sSelectorToCalico(&peer.Pods.PodSelector, SelectorPod)
+				nsSelector = c.k8sSelectorToCalico(&peer.Pods.NamespaceSelector, SelectorNamespace)
+			}
+			// Build inbound rule and append to list.
+			rules = append(rules, apiv3.Rule{
+				Action:   apiv3.Action(rule.Action),
+				Name:     rule.Name,
+				Protocol: protocol,
+				Source: apiv3.EntityRule{
+					Selector:          selector,
+					NamespaceSelector: nsSelector,
+				},
+				Destination: apiv3.EntityRule{
+					Ports: calicoPorts,
+				},
+			})
+		}
+	}
+
+	return rules, nil
+}
+
+func (c converter) k8sANPEgressRuleToCalico(rule adminpolicy.AdminNetworkPolicyEgressRule) ([]apiv3.Rule, error) {
+	rules := []apiv3.Rule{}
+	peers := []*adminpolicy.AdminNetworkPolicyEgressPeer{}
+	ports := []*adminpolicy.AdminNetworkPolicyPort{}
+
+	// Built up a list of the sources and a list of the destinations.
+	for _, r := range rule.To {
+		// We need to add a copy of the peer so all the rules don't
+		// point to the same location.
+		peer := &adminpolicy.AdminNetworkPolicyEgressPeer{Namespaces: r.Namespaces}
+		if r.Namespaces == nil && r.Pods != nil {
+			peer.Pods = &adminpolicy.NamespacedPod{
+				NamespaceSelector: r.Pods.NamespaceSelector,
+				PodSelector:       r.Pods.PodSelector,
+			}
+		}
+		peers = append(peers, peer)
+	}
+
+	if rule.Ports != nil {
+		for _, p := range *rule.Ports {
+			// We need to add a copy of the port so all the rules don't
+			// point to the same location.
+			port := adminpolicy.AdminNetworkPolicyPort{}
+			if p.PortNumber != nil {
+				port.PortNumber = &adminpolicy.Port{
+					Protocol: ensureProtocol(p.PortNumber.Protocol),
+					Port:     p.PortNumber.Port,
+				}
+			}
+			if p.PortRange != nil {
+				port.PortRange = &adminpolicy.PortRange{
+					Protocol: ensureProtocol(p.PortRange.Protocol),
+					Start:    p.PortRange.Start,
+					End:      p.PortRange.End,
+				}
+			}
+			ports = append(ports, &port)
+		}
+	}
+
+	// If there no peers, or no ports, represent that as nil.
```

Comment:
I think the spec requires there to be at least one peer (although that means it's hard to do basic things like allow all/deny all!).  I filed an issue upstream for that: https://github.com/kubernetes-sigs/network-policy-api/issues/248